### PR TITLE
CNFCERT-322:add test for checking image tag exists]

### PR DESCRIPTION
**Pull request by aabughosh**

### CODE REVIEW

File: CATALOG.md [hunk 1]
```diff
@@ -481,6 +481,19 @@ Result Type|normative
...
+
+Property|Description
+---|---
+Test Case Name|test-Containers-Image-Tag
+Test Case Label|lifecycle-test-Containers-Image-Tag
+Unique ID|http://test-network-function.com/testcases/lifecycle/test-Containers-Image-Tag
+Version|v1.0.0
+Description|http://test-network-function.com/testcases/lifecycle/test-Containers-Image-Tag Check that image tag exists on containers.
+Result Type|informative
+Suggested Remediation|Ensure that all persistent volumes are using the reclaim policy: delete
```
Comment (sebrandon1) [inline CATALOG.md:494]: This needs its own remediation string.

File: cnf-certification-test/identifiers/identifiers.go [hunk 2]
```diff
@@ -362,6 +362,10 @@ var (
 		Url:     formTestURL(common.LifecycleTestKey, "persistent-volume-reclaim-policy"),
 		Version: versionOne,
 	}
+	TestContainersImageTag = claim.Identifier{
+		Url:     formTestURL(common.LifecycleTestKey, "test-Containers-Image-Tag"),
```
Comment (sebrandon1) [inline cnf-certification-test/identifiers/identifiers.go:366]:
```suggestion
		Url:     formTestURL(common.LifecycleTestKey, "test-Containers-image-Tag"),
```

No need for capitalization.

File: cnf-certification-test/identifiers/identifiers.go [hunk 3]
```diff
@@ -993,6 +997,14 @@ that there are no changes to the following directories:
 		BestPracticeReference: bestPracticeDocV1dot4URL + " Section 3.3.4",
 		ExceptionProcess:      NoDocumentedProcess,
 	},
+	TestContainersImageTag: {
+		Identifier:            TestContainersImageTag,
+		Type:                  informativeResult,
+		Description:           formDescription(TestContainersImageTag, `Check that image tag exists on containers.`),
+		Remediation:           PersistentVolumeReclaimPolicyRemediation,
```
Comment (sebrandon1) [inline cnf-certification-test/identifiers/identifiers.go:1004]: Needs own remediation var.

File: cnf-certification-test/lifecycle/suite.go [hunk 4]
```diff
@@ -147,6 +148,12 @@ var _ = ginkgo.Describe(common.LifecycleTestKey, func() {
 	ginkgo.It(testID, ginkgo.Label(testID), func() {
 		testPodPersistentVolumeReclaimPolicy(&env)
 	})
+
+	testID = identifiers.XformToGinkgoItIdentifier(identifiers.TestContainersImageTag)
+	ginkgo.It(testID, ginkgo.Label(testID), func() {
+		testContainersImageTag(&env)
```
Comment (sebrandon1) [inline cnf-certification-test/lifecycle/suite.go:154]: Add `testhelper.SkipIfEmptyAny(ginkgo.Skip, env.Containers)` before the call to the func.

File: cnf-certification-test/lifecycle/suite.go [hunk 5]
```diff
@@ -483,3 +490,19 @@ func testPodPersistentVolumeReclaimPolicy(env *provider.TestEnvironment) {
 		ginkgo.Fail(errMsg)
 	}
 }
+func testContainersImageTag(env *provider.TestEnvironment) {
+	testhelper.SkipIfEmptyAll(ginkgo.Skip, env.Containers)
```
Comment (sebrandon1) [inline cnf-certification-test/lifecycle/suite.go:494]: Move this to the ginkgo.It block.

File: cnf-certification-test/lifecycle/suite.go [hunk 6]
```diff
@@ -483,3 +490,19 @@ func testPodPersistentVolumeReclaimPolicy(env *provider.TestEnvironment) {
...
+		logrus.Debugln("check container ", cut.String(), " image, should be tagged ")
+		image := strings.Split(cut.Data.Image, ":")
+		if len(image) <= 1 {
+			badcontainers = append(badcontainers, "{"+cut.String()+": is not having tag"+"}")
+			logrus.Errorln("container image", cut.Data.Name, " is not having tag ")
+		}
+	}
+	if len(badcontainers) > 0 {
+		tnf.ClaimFilePrintf("Containers have been found that missing tag on image: %v", badcontainers)
+		ginkgo.Fail("Containers have been found that missing tag on image.")
```
Comment (sebrandon1) [inline cnf-certification-test/lifecycle/suite.go:506]:
```suggestion
		ginkgo.Fail("Containers have been found that are missing tag(s) on image.")
```

File: cnf-certification-test/lifecycle/suite.go [hunk 7]
```diff
@@ -483,3 +490,19 @@ func testPodPersistentVolumeReclaimPolicy(env *provider.TestEnvironment) {
...
 }
+func testContainersImageTag(env *provider.TestEnvironment) {
+	testhelper.SkipIfEmptyAll(ginkgo.Skip, env.Containers)
+	badcontainers := []string{}
+	for _, cut := range env.Containers {
+		logrus.Debugln("check container ", cut.String(), " image, should be tagged ")
+		image := strings.Split(cut.Data.Image, ":")
+		if len(image) <= 1 {
+			badcontainers = append(badcontainers, "{"+cut.String()+": is not having tag"+"}")
+			logrus.Errorln("container image", cut.Data.Name, " is not having tag ")
```
Comment (sebrandon1) [inline cnf-certification-test/lifecycle/suite.go:501]:
```suggestion
			logrus.Debugf("Container image for %s does not have a valid tag", cut.Data.Name)
```

**Review by sebrandon1 — COMMENTED** [review]
Just a couple of comments

File: cnf-certification-test/identifiers/remediation.go [hunk 8]
```diff
@@ -182,5 +182,7 @@ const (
 
 	PersistentVolumeReclaimPolicyRemediation = `Ensure that all persistent volumes are using the reclaim policy: delete`
 
+	ContainersImageTag = `Ensure that all the container iamge are tagged`
```
Comment (sebrandon1) [inline cnf-certification-test/identifiers/remediation.go:185]:
```suggestion
	ContainersImageTag = `Ensure that all the container images are tagged`
```

**Review by sebrandon1 — COMMENTED** [review]
One more comment!

**Comment by sebrandon1** [comment]
@edcdavid can you look at the smoke tests here?  It seems like its failing due to something from #424

**Review by edcdavid — COMMENTED** [review]
Fix for failing CI at 
https://github.com/test-network-function/cnf-certification-test-partner/pull/151

**Review by sebrandon1 — APPROVED** [review]

File: CATALOG.md [hunk 9]
```diff
@@ -481,6 +481,19 @@ Result Type|normative
 Suggested Remediation|Ensure CNF statefulsets/replica sets can scale in/out successfully.
 Best Practice Reference|https://connect.redhat.com/sites/default/files/2022-05/Cloud%20Native%20Network%20Function%20Requirements%201-3.pdf Section 5.2
 Exception Process|There is no documented exception process for this.
+#### test-containers-image-tag
+
+Property|Description
+---|---
+Test Case Name|test-containers-image-tag
+Test Case Label|lifecycle-test-containers-image-tag
```
Comment (edcdavid) [inline CATALOG.md:489]: Shouldn't this test be in the affiliated-certification suite?

File: CATALOG.md [hunk 10]
```diff
@@ -481,6 +481,19 @@ Result Type|normative
...
+Unique ID|http://test-network-function.com/testcases/lifecycle/test-containers-image-tag
+Version|v1.0.0
+Description|http://test-network-function.com/testcases/lifecycle/test-containers-image-tag Check that image tag exists on containers.
+Result Type|informative
+Suggested Remediation|Ensure that all the container iamge are tagged
```
Comment (edcdavid) [inline CATALOG.md:494]: iamge -> image

Comment (sebrandon1) [reply CATALOG.md:494]: I think all that needs to be done is `make build-catalog-md`.

File: CATALOG.md (hunk 9 above)
Comment (aabughosh) [reply CATALOG.md:489]: we have the image pull policy in the lifecycle so its similar to it

Comment (jmontesi) [reply CATALOG.md:489]: Agree. Lifecycle does not seem to be the place for this test.

File: cnf-certification-test/identifiers/identifiers.go [hunk 11]
```diff
@@ -362,6 +362,10 @@ var (
 		Url:     formTestURL(common.LifecycleTestKey, "persistent-volume-reclaim-policy"),
 		Version: versionOne,
 	}
+	TestContainersImageTag = claim.Identifier{
+		Url:     formTestURL(common.LifecycleTestKey, "test-containers-image-tag"),
```
Comment (jmontesi) [inline cnf-certification-test/identifiers/identifiers.go:366]: I have not seen the prefix "test" often in our suite. I'd use "containers-image-tag"

File: cnf-certification-test/lifecycle/suite.go [hunk 12]
```diff
@@ -483,3 +490,17 @@ func testPodPersistentVolumeReclaimPolicy(env *provider.TestEnvironment) {
 		ginkgo.Fail(errMsg)
 	}
 }
+func testContainersImageTag(env *provider.TestEnvironment) {
+	badcontainers := []string{}
+	for _, cut := range env.Containers {
+		logrus.Debugln("check container ", cut.String(), " image, should be tagged ")
+		if cut.ContainerImageIdentifier.Tag == "" {
+			badcontainers = append(badcontainers, "{"+cut.String()+": is not having tag"+"}")
```
Comment (jmontesi) [inline cnf-certification-test/lifecycle/suite.go:498]: That's a strange format. I'd stick to just adding the cut.String() to the slice. After all, when printing the log messages you are already stating that they refer to containers not having their images tagged.

File: cnf-certification-test/lifecycle/suite.go [hunk 13]
```diff
@@ -483,3 +490,17 @@ func testPodPersistentVolumeReclaimPolicy(env *provider.TestEnvironment) {
...
+			logrus.Debugf("Container image for %s does not have a valid tag", cut.Data.Name)
```
Comment (jmontesi) [inline cnf-certification-test/lifecycle/suite.go:499]: Is this log is necessary?

**Review by jmontesi — APPROVED** [review]
Approved pending fix for typos.

**Pull request merged**